luda-adventure/src/app/game/types/game_object/types/positioner/predict_collision.rs

```diff
@@ -0,0 +1,168 @@
+        .iter()
+        .filter_map(|target_collision_box| {
+            predict_collision_between_character_collision_box_and_others(
+                &character_collision_box,
+                target_collision_box,
+                velocity,
+                current_time,
+                end_time,
+                start_position,
+            )
+        })
+        .filter(|collision_prediction| collision_prediction.start_time >= 0.0.ms())
+        .collect::<Vec<_>>();
+    collision_prediction_list.sort_by(|a, b| a.start_time.cmp(&b.start_time));
+    collision_prediction_list.into_iter().next()
+}
+
+fn predict_collision_between_character_collision_box_and_others(
+    character_collision_box: &CollisionBox,
+    target_collision_box: &CollisionBox,
+    velocity: Velocity,
+    start_time: Time,
+    end_time: Time,
+    start_position: Xy<Tile>,
+) -> Option<CollisionPrediction> {
+    let intersection_prediction_of_x = predict_intersection_on_one_axis(
+        velocity.x,
+        character_collision_box.left(),
+        character_collision_box.right(),
+        target_collision_box.left(),
+        target_collision_box.right(),
+    );
+    let intersection_prediction_of_y = predict_intersection_on_one_axis(
+        velocity.y,
+        character_collision_box.top(),
+        character_collision_box.bottom(),
+        target_collision_box.top(),
+        target_collision_box.bottom(),
+    );
+    match (intersection_prediction_of_x, intersection_prediction_of_y) {
+        (
+            IntersectionPrediction::WillIntersect {
+                start_remaining_time: intersection_start_remaining_time_of_x,
+                end_remaining_time: intersection_end_remaining_time_of_x,
+                start_remaining_distance: intersection_start_remaining_distance_of_x,
+            },
+            IntersectionPrediction::WillIntersect {
+                start_remaining_time: intersection_start_remaining_time_of_y,
+                end_remaining_time: intersection_end_remaining_time_of_y,
+                start_remaining_distance: intersection_start_remaining_distance_of_y,
+            },
+        ) => {
+            let collision_direction = match intersection_start_remaining_time_of_x
+                > intersection_start_remaining_time_of_y
+            {
+                true => CollisionDirection::Horizontal,
+                false => CollisionDirection::Vertical,
+            };
+            let (collision_start_remaining_time, collision_end_remaining_time) =
+                match collision_direction {
+                    CollisionDirection::Horizontal => (
+                        intersection_start_remaining_time_of_x,
+                        intersection_end_remaining_time_of_y,
+                    ),
+                    CollisionDirection::Vertical => (
+                        intersection_start_remaining_time_of_y,
+                        intersection_end_remaining_time_of_x,
```

Comment:
is it right that `collision_end_remaining_time` is end of x or y? not the `start` of counter-side?

Reply:
Here is a design sketch for this.

![image](https://user-images.githubusercontent.com/38313680/194223061-e4aa53ad-87d9-4ff7-8256-1b634a81d979.png)